src/pybind/mgr/volumes/fs_subvolume.py

```diff
@@ -0,0 +1,365 @@
+        for an already-created subvolume, even if it is in use.
+
+        :param subvolume_path: SubvolumePath instance
+        :param size: In bytes, or None for no size limit
+        :param namespace_isolated: If true, use separate RADOS namespace for this subvolume
+        :return: None
+        """
+        path = self._subvolume_path(subvolume_path)
+        log.info("creating subvolume with path: {0}".format(path))
+
+        self._mkdir_p(path, mode)
+
+        if size is not None:
+            self.fs.setxattr(path, 'ceph.quota.max_bytes', to_bytes(size), 0)
+
+        # enforce security isolation, use separate namespace for this subvolume
+        if namespace_isolated:
+            namespace = "{0}{1}".format(self.pool_ns_prefix, subvolume_path.subvolume_id)
+            log.info("creating subvolume with path: {0}, using rados namespace {1} to isolate data.".format(subvolume_path, namespace))
+            self.fs.setxattr(path, 'ceph.dir.layout.pool_namespace',
+                             to_bytes(namespace), 0)
+        else:
+            # If subvolume's namespace layout is not set, then the subvolume's pool
+            # layout remains unset and will undesirably change with ancestor's
+            # pool layout changes.
+            pool_name = self._get_ancestor_xattr(path, "ceph.dir.layout.pool")
+            self.fs.setxattr(path, 'ceph.dir.layout.pool',
+                             to_bytes(pool_name), 0)
+
+    def delete_subvolume(self, subvolume_path):
+        """
+        Make a subvolume inaccessible to guests.  This function is idempotent.
+        This is the fast part of tearing down a subvolume: you must also later
+        call purge_subvolume, which is the slow part.
+
+        :param subvolume_path: Same identifier used in create_subvolume
+        :return: None
+        """
+
+        path = self._subvolume_path(subvolume_path)
+        log.info("deleting subvolume with path: {0}".format(path))
+
+        # Create the trash folder if it doesn't already exist
+        trash = os.path.join(self.subvolume_prefix, "_deleting")
+        self._mkdir_p(trash)
+
+        # We'll move it to here
+        trashed_subvolume = os.path.join(trash, subvolume_path.subvolume_id)
+
+        # Move the subvolume to the trash folder
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            log.warning("Trying to delete subvolume with path '{0}' but it's already gone".format(
+                path))
+            return
+        else:
+            self.fs.rename(path, trashed_subvolume)
+
+    def purge_subvolume(self, subvolume_path):
+        """
+        Finish clearing up a subvolume that was previously passed to delete_subvolume.  This
+        function is idempotent.
+        """
+
+        trash = os.path.join(self.subvolume_prefix, "_deleting")
+        trashed_subvolume = os.path.join(trash, subvolume_path.subvolume_id)
+
+        try:
+            self.fs.stat(trashed_subvolume)
```

Comment:
nit: we could probably skip this `stat` and handle the exception in `opendir()` (in `rmtree()`)?

(and we need to do that probably since the dir can go away b/w `stat` and `opendir`... accidental deletion)

Reply:
Fixed.